warehouse/tuf/services.py

```diff
@@ -0,0 +1,414 @@
+            )
+        else:
+            return datetime.datetime.now().replace(microsecond=0) + datetime.timedelta(
+                seconds=self._request.registry.settings[f"tuf.{role_name}.expiry"]
+            )
+
+    def init_repository(self):
+        """
+        Creates TUF top-level role metadata (root, targets, snapshot, timestamp).
+
+        The metadata is populated with configured expiration times, signature thresholds
+        and verification keys, and signed and persisted using the configured key and
+        storage services.
+
+        FIXME: In production 'root' and 'targets' roles require offline singing keys,
+        which may not be available at the time of initializing this metadata.
+        """
+        metadata_repository = MetadataRepository(
+            self._storage_backend, self._key_storage_backend
+        )
+
+        if metadata_repository.is_initialized:
+            raise FileExistsError("TUF Metadata Repository files already exists.")
+
+        top_roles_payload = dict()
+        for role in TOP_LEVEL_ROLE_NAMES:
+            top_roles_payload[role] = RolesPayload(
+                expiration=self._set_expiration_for_role(role),
+                threshold=self._request.registry.settings[f"tuf.{role}.threshold"],
+                keys=self._key_storage_backend.get(role, "private"),
+            )
+
+        metadata_repository.initialize(top_roles_payload, True)
+
+    def init_targets_delegation(self):
+        """
+        Creates TUF metadata for hash bin delegated targets roles (bins, bin-n).
+
+        Metadata is created for one 'bins' role and a configured number of 'bin-n'
+        roles. It is populated with configured expiration times, signature thresholds
+        and verification keys, and signed and persisted using the configured key and
+        storage services.
+
+        FIXME: In production the 'bins' role requires an offline singing key, which may
+        not be available at the time of initializing this metadata.
+        """
+        hash_bins = self._get_hash_bins()
+        metadata_repository = MetadataRepository(
+            self._storage_backend, self._key_storage_backend
+        )
+
+        # Top-level 'targets' role delegates trust for all target files to 'bins' role.
+        delegate_roles_payload = dict()
+        delegate_roles_payload["targets"] = list()
+        delegate_roles_payload["targets"].append(
+            RolesPayload(
+                expiration=self._set_expiration_for_role(Role.BINS.value),
+                threshold=self._request.registry.settings[
+                    f"tuf.{Role.BINS.value}.threshold"
+                ],
+                keys=self._key_storage_backend.get(Role.BINS.value, "private"),
+                delegation_role=Role.BINS.value,
+                paths=["*", "*/*/*/*"],
+            )
+        )
+
+        snapshot_metadata = metadata_repository.delegate_targets_roles(
+            delegate_roles_payload,
+        )
```

Comment:
From a TUF perspective it is enough to bump snapshot only once after having updated all targets metadata (targets, bins, bin-n). And from a performance perspective it would be a good idea too.

I guess the reason why you bump snapshot here is to write the new snapshot to storage, so that `delegate_targets_roles` below, which loads snapshot from storage doesn't have the old version. Is that correct?

Reply:
What would be the benefit of multiple versions?

Reply:
Not really. I just did it to test it and see if my steps were correct.

Reply:
I suggest to revert it then.

Reply:
I just realized that `delegate_targets_roles` takes a dictionary of `{"<delegator>": ["<delegatee>", ...]}`. So you could actually build your `delegate_roles_payload` with "targets" and "bins" and call `delegate_targets_roles` only once (and bump_snapshot) afterwards.